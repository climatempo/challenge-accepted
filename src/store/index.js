import { createStore, applyMiddleware } from 'redux';
import ReduxThunk from 'redux-thunk';
import reducers from './reducers';

const devTools = window.__REDUX_DEVTOOLS_EXTENSION__ && window.__REDUX_DEVTOOLS_EXTENSION__();

const store = applyMiddleware(ReduxThunk)(createStore)(reducers, devTools);

export default store;
