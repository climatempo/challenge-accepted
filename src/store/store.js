import { createStore, applyMiddleware, compose } from "redux";
import thunk from 'redux-thunk';

import reducer from "../reducer/reducer";

export default createStore(reducer, compose(applyMiddleware(thunk)));