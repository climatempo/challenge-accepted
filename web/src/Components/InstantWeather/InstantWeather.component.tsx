import { useAppSelector } from '../../hooks';
import { ReactNode, useEffect, useState } from 'react';
import { CircularProgress, Typography } from '@mui/material';
import { useSnackbar } from 'notistack';

type InstantWeather = {
    idlocale: number,
    temperature: number,
    icon: string,
    condition: string,
    humidity: number,
    sensation: number,
    windVelocity: number,
    pressure: number,
    date: string
}

function InstantWeather() {
    const locale = useAppSelector(state => state.locale.value);
    const tempUnity = useAppSelector(state => state.temperature.value);
    const [weather, setWeather] = useState<InstantWeather>();
    const [loading, setLoading] = useState<boolean>(true);
    const { enqueueSnackbar: notify } = useSnackbar();
    let fetchBlock = false;

    useEffect(() => {
        if(!locale)
            return;
        setLoading(true);
        fetch(`http://localhost/city/${ locale.idlocale }/now`)
            .then(response => {
                setLoading(false);
                if(response.status === 404) {
                    notify('Não temos o clima dessa cidade.', { variant: 'info' });
                    return undefined;
                }else return response.json();
            })
            .then(weather => setWeather(weather))
            .catch(err => {
                fetchBlock = true;
                setLoading(false);
                console.error(err);
                notify('Alguma coisa deu errado.', { variant: 'error'});
                setTimeout(() => fetchBlock = false, 2000);
            });
    }, [locale]);

    function renderSubInfo(caption: string, info: string|number, unity: ReactNode) {
        return <div className="text-center sm:text-left">
            <div className="hidden sm:flex">
                <Typography variant='h4' className="w-fit">
                    { info }
                </Typography>
                <Typography variant='subtitle1' className="self-end w-fit">
                    { unity }
                </Typography>
            </div>
            <div className="flex sm:hidden">
                <Typography variant='h5' className="w-fit">
                    { info }
                </Typography>
                <Typography variant='subtitle1'>
                    { unity }
                </Typography>
            </div>
            <Typography variant='caption'>
                { caption }
            </Typography>
        </div>;
    }

    function renderContent(weather: InstantWeather) {
        const { temperature, windVelocity, pressure, sensation, humidity, condition } = weather;
        return <>
            <div className="flex flex-col sm:flex-row items-center sm:items-end gap-x-4">
                <Typography variant='h1'>
                    { tempUnity === 'c' ? temperature + '°C' : ((temperature * 1.8) + 32).toFixed(0) + '°F' }
                </Typography>
                <Typography variant='h4' className="lg:hidden pb-3">
                    { condition }
                </Typography>
            </div>
            <div className="flex flex-col gap-3 w-full">
                <Typography variant='h6' className="hidden lg:block">
                    { condition }
                </Typography>
                <div className="grid grid-cols-2 sm:grid-cols-4 gap-8 max-w-[400px] w-full sm:max-w-full mx-auto">
                    { renderSubInfo('Sensação', tempUnity === 'c'
                        ? sensation
                        : ((sensation * 1.8) + 32).toFixed(0), tempUnity === 'c' ?
                        <div className="mt-[-2.5rem]">°C</div>
                        : <div className="mt-[-2.5rem]">°F'</div>) }
                    { renderSubInfo('Humidade', humidity, <div className="mt-[-2.5rem]">%</div>)}
                    { renderSubInfo('Vento', tempUnity === 'c'
                        ? windVelocity
                        : (windVelocity/1.609).toFixed(0), tempUnity === 'c'
                        ? <div className="">km/h</div>
                        : <div className="">mph</div>) }
                    { renderSubInfo('Pressão', pressure, <div className="">hPa</div>) }
                </div>
            </div>
        </>;
    }

    function renderSkeleton() {
        return loading ? <CircularProgress className="mx-auto" /> : <></>
    }

    return <div className="flex gap-10 items-center max-w-[720px] mx-auto flex-col lg:flex-row">
        { weather && !loading ? renderContent(weather) : renderSkeleton() }
    </div>;
}

export default InstantWeather;