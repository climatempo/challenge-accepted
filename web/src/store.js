import Vue from "vue";
import Vuex from "vuex";
import localeStore from "./store/locale";
import weatherStore from "./store/weather";

Vue.use(Vuex);

export default new Vuex.Store({
  state: {},
  mutations: {},
  actions: {},
  modules: [
      localeStore,
      weatherStore
  ]
});
