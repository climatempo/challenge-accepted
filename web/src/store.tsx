import { configureStore } from '@reduxjs/toolkit';
import themeReducer from './Reducers/Theme.reducer';
import temperatureReducer from './Reducers/Temperature.reducer';
import precipitationReducer from './Reducers/Precipitation.reducer';

const store = configureStore({
    reducer: {
        theme: themeReducer,
        temperature: temperatureReducer,
        precipitation: precipitationReducer,
    }
});

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;
export default store;