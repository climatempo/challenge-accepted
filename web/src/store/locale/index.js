import mutations from './mutations'
import actions from './actions'
import getters from './getters'

/**
 * Store de localizaçeõs
 *
 * @type mixins
 */
const localeStore = {
    state: {
        locales: []
    },
    mutations: mutations,
    actions: actions,
    getters: getters
};

export default localeStore
