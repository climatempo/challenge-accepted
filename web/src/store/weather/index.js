import mutations from './mutations'
import actions from './actions'
import getters from './getters'

/**
 * Store de previs√µes
 *
 * @type mixins
 */
const weatherStore = {
    state: {
        weather: []
    },
    mutations: mutations,
    actions: actions,
    getters: getters
};

export default weatherStore
