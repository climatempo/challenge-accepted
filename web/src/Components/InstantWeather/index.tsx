import InstantWeather from './InstantWeather.component';

export default InstantWeather;