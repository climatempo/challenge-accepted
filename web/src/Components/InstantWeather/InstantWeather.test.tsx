import renderer from 'react-test-renderer';
import './../../Tests/watchMedia.mock';
import { Provider } from 'react-redux';
import store from '../../store';
import { SnackbarProvider } from 'notistack';
import InstantWeather from './index';

test("Instant WeatherController component", () => {
  const reduxStore = store;

  const weather = renderer.create(
      <Provider store={reduxStore}>
        <SnackbarProvider maxSnack={4}>
          <InstantWeather/>
        </SnackbarProvider>
      </Provider>
  );

  expect(weather.toJSON()).toMatchSnapshot();
});