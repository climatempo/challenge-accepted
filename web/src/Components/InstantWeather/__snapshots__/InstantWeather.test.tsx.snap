// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Instant WeatherController component 1`] = `
<div
  className="flex gap-10 items-center max-w-[720px] mx-auto flex-col lg:flex-row"
>
  <span
    className="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary mx-auto css-18lrjg1-MuiCircularProgress-root"
    role="progressbar"
    style={
      Object {
        "height": 40,
        "width": 40,
      }
    }
  >
    <svg
      className="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
      viewBox="22 22 44 44"
    >
      <circle
        className="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
        cx={44}
        cy={44}
        fill="none"
        r={20.2}
        strokeWidth={3.6}
        style={Object {}}
      />
    </svg>
  </span>
</div>
`;
